jsign-crypto/src/main/java/net/jsign/jca/HashiCorpVaultSigningService.java

```diff
@@ -131,15 +198,81 @@ public byte[] sign(SigningServicePrivateKey privateKey, String algorithm, byte[]
 
         Map<String, Object> request = new HashMap<>();
         request.put("key_version", keyVersion);
-        request.put("digest", Base64.getEncoder().encodeToString(data));
+
+        if (this.googleCloudKMS) {
+            request.put("digest", Base64.getEncoder().encodeToString(data));
+        }
+        else {
+            String shaType;
+            try {
+                shaType = digestAlgorithm.toString().split("SHA")[1];
+            } catch (RuntimeException e) {
+                logger.log(Level.WARNING, "Could not determine SHA type - defaulting to SHA256");
+                shaType = "256";
+            }
+
+            request.put("input", Base64.getEncoder().encodeToString(data));
+            request.put("prehashed", true);
+            request.put("hash_algorithm", String.format("sha2-%s", shaType));
+
+            // By default, RSA sign in HashiCorp Vault Transit uses RSA-PSS.
+            if (privateKey.getAlgorithm().equals("RSA")) {
+                request.put("signature_algorithm", "pkcs1v15");
+            }
+        }
 
         try {
             Map<String, ?> response = client.post("sign/" + keyName, JsonWriter.format(request));
-            String signature = ((Map<String, String>) response.get("data")).get("signature");
+
+            if (response == null || !response.containsKey("data")) {
+                throw new IllegalArgumentException("Response or 'data' key is null.");
+            }
+
+            if (response.containsKey("warnings") && response.get("warnings") != null) {
+                logger.log(Level.WARNING, String.format("Vault responded with warnings: %s", response.get("warnings")));
+            }
+
+            if (response.containsKey("error")) {
+                throw new IllegalArgumentException(String.format("Vault responded with error: %s", response.get("error")));
```

Comment:
Are you sure this statement is reached? If the server returns an HTTP error 4xx or 5xx it will be caught by the error handling code in RESTClient.

Reply:
You are correct. This code would otherwise not be reachable. 
Removed in next commit.